internal/pubsub/samples/sub/gcp.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	raw "cloud.google.com/go/pubsub/apiv1"
+	"gocloud.dev/gcp"
+	"gocloud.dev/internal/pubsub"
+	"gocloud.dev/internal/pubsub/gcppubsub"
+	"google.golang.org/api/option"
+	"google.golang.org/grpc"
+	grpccreds "google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+)
+
+// openGCPSubscription returns the GCP topic based on the subscription ID.
+func openGCPSubscription(ctx context.Context, subID string) (*pubsub.Subscription, func(), error) {
+	subClient, cleanup, err := openGCPSubscriberClient(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	proj, err := getGCPProjectID()
+	if err != nil {
+		return nil, nil, err
+	}
+	sub := gcppubsub.OpenSubscription(ctx, subClient, proj, subID, nil)
+	cleanup2 := func() {
+		sub.Shutdown(ctx)
+		cleanup()
+	}
+	return sub, cleanup2, nil
+}
+
+func getGCPProjectID() (gcp.ProjectID, error) {
+	p := os.Getenv("PROJECT_ID")
+	if p == "" {
+		return gcp.ProjectID(""), errors.New("$PROJECT_ID should be set to the GCP project")
+	}
+	return gcp.ProjectID(p), nil
+}
+
+// openGCPSubscriberClient opens a GCP SubscriberClient with default credentials.
+func openGCPSubscriberClient(ctx context.Context) (*raw.SubscriberClient, func(), error) {
+	creds, err := gcp.DefaultCredentials(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting default credentials: %v", err)
+	}
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(grpccreds.NewClientTLSFromCert(nil, "")),
+		grpc.WithPerRPCCredentials(oauth.TokenSource{TokenSource: gcp.CredentialsTokenSource(creds)}),
+	}
+	conn, err := grpc.DialContext(ctx, gcppubsub.EndPoint, opts...)
```

Comment:
Seems like there's a fair amount of shared code between this and the `pub` command. Would joining them into a single binary make sense?

Reply:
Done.